### ISSUE TITLE
DOM lib: Add support for Trusted Types API

### ISSUE THREAD

**Issue by koto** [Microsoft/TypeScript#30024]
## Search Terms

Trusted Types, DOM 

## Suggestion

[Trusted Types](https://github.com/WICG/trusted-types) ([spec](https://wicg.github.io/trusted-types/dist/spec/), [introductory article](https://developers.google.com/web/updates/2019/02/trusted-types)) is a new experimental DOM API implemented within the WICG , with a working [Chrome implementation](https://www.chromestatus.com/feature/5650088592408576). 

The API creates a few new objects available on the global object in the browser, like most other web APIs ([impl in TS](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/trusted-types/index.d.ts) and [in Closure compiler](https://github.com/WICG/trusted-types/blob/master/externs/externs.js)).

Under certain conditions, controlled by a HTTP header (analogous to Content-Security-Policy behavior), the API can enable the *enforcement* - then it changes the signature of several DOM API functions and property setters, such that they accept specific object types, and reject strings. Colloquially, DOM API becomes strongly typed. 

For example, with Trusted Types `Element.innerHTML` property setter accepts a `TrustedHTML` object.

Trusted Type objects stringify to their inner value. This API shape is a deliberate choice that enables existing web applications and libraries to gradually migrate from strings to Trusted Types without breaking functionality. In our example, it makes it possible to write the following:

```js
const policy = TrustedTypes.createPolicy('foo', { 
  createHTML: (s) => { /* some validation*/; return s} 
});

const trustedHTML = policy.createHTML('bar');
anElement.innerHTML = trustedHTML

anElement.innerHTML === 'bar'
```
The above code work regardless if the Trusted Types *enforcement* is enabled or not. 

Reading from the DOM is unaffected, so `Element.innerHTML` getter returns a string. That's for practical reasons -- web applications read from DOM more often than they write to it, and only writing exposes the application to DOM XSS risks. Typing only the setters allows s to secure web applications with minimal code changes.

It's difficult to map that API using TS types (due to #2521). The only way is to change the DOM sink functions to have the `any` type, which is obviously suboptimal. 

## Use Cases

Writing a TS application that uses the Trusted Types API.

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by engelsdamien** [comment]
For an additional data point, the way the closure compiler supports those sink assignments seems to be through the [@implicitCast](https://github.com/google/closure-compiler/wiki/Annotating-JavaScript-for-the-Closure-Compiler#implicitcast) annotation.

**Comment by koto** [comment]
Trusted Types are now available in Chrome canary.

**Comment by wibimaster** [comment]
Need it, it's on Chrome official release :x

**Comment by shicks** [comment]
FWIW, it looks like there's finally some traction on #2521 so this might become feasible at some point.

**Comment by shhnjk** [comment]
@RyanCavanaugh, any update on this?

**Comment by koto** [comment]
The setters/getters typing needed to unblock this will, to my understanding, ship in 4.3. Should we create a PR to change the `lib.dom.d.ts` type annotation for the sinks to align with https://w3c.github.io/webappsec-trusted-types/dist/spec/ ? Essentially, in a couple of places in DOM the setter type would be `TrustedHTML | string` (or similar unions for other types), instead of `string`.

**Comment by aryzing** [comment]
Seems to we well supported now

* https://developer.mozilla.org/en-US/docs/Web/API/TrustedTypePolicyFactory/createPolicy

Would be great to have TS support for this.

**Comment by shicks** [comment]
The necessary TS language support landed in 4.3: https://devblogs.microsoft.com/typescript/announcing-typescript-4-3/#separate-write-types

I had assumed somebody was updating the standard library declarations to accommodate this, but maybe that never happened?  In any case, as far as I'm aware that's all that's left to do, and it should be pretty straightforward.

**Comment by shhnjk** [comment]
> Should we create a PR to change the `lib.dom.d.ts` type annotation for the sinks to align with https://w3c.github.io/webappsec-trusted-types/dist/spec/ ? Essentially, in a couple of places in DOM the setter type would be `TrustedHTML | string` (or similar unions for other types), instead of `string`.

Yes, if you have bandwidth @koto! :)

**Comment by tosmolka** [comment]
This is starting to block our adoption of Trusted Types so we took a look as well. Is https://github.com/microsoft/TypeScript-DOM-lib-generator/pull/1246 roughly what you guys had in mind as next step? Can you pls take a look? Thanks.

**Comment by shicks** [comment]
microsoft/TypeScript-DOM-lib-generator#1246 is backward-incompatible and will break a bunch of existing code.  The fix is to only widen the type on writing, while leaving the read type the same:

```
interface Element {
  get outerHTML(): string;
  set outerHTML(html: string|TrustedHTML);
}
```

**Comment by tosmolka** [comment]
@shicks , the PR with the fix has been updated, can you please check the fix again for backward-compatibility? Thanks.

**Comment by freshp86** [comment]
Any updates here?

**Comment by HolgerJeromin** [comment]
> Any updates here?

No support from two engines (not browsers), [no ts lib support](https://github.com/microsoft/TypeScript-DOM-lib-generator?tab=readme-ov-file#why-is-my-fancy-api-still-not-available-here)...

**Comment by freshp86** [comment]
> No support from two engines (not browsers), [no ts lib support](https://github.com/microsoft/TypeScript-DOM-lib-generator?tab=readme-ov-file#why-is-my-fancy-api-still-not-available-here)...

Firefox official response: https://github.com/mozilla/standards-positions/issues/20#issuecomment-1853427823
Firefox implementation tracking bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1508286

Current status according to https://caniuse.com/mdn-api_trustedtypes
_"Can be enabled by setting `dom.security.trusted_types.enabled` to true"_

It's not clear from the link above if "supported" means "enableld by deafult". There are enough signals that this will be available on Firefox by default soon, and already available behind a flag (verified locally as well on v137.0.1).
Thanks!

**Comment by koto** [comment]
FWIW it's also enabled in Safari Tech Preview 215 - https://webkit.org/blog/16523/release-notes-for-safari-technology-preview-215/

**Comment by HolgerJeromin** [comment]
So no one need to inform or ping anyone. 

The browsers will publish the feature.
Someone will add this to the mdn github repo with the compatibility. 
From that the tables in mdn are generated but also the lib d.ts files from typescript.
A bot will create a PR for the change to typescript repo.
After merge this will be included in next ts version.

**Comment by hybrist** [comment]
Adding some quick research notes here:

1. As of Sept 2025, there are two shipping browser engines that include Trusted Types (Safari + Chrome).
2. This is also reflected on MDN and in the list of features in `mdn.json` in https://github.com/microsoft/TypeScript-DOM-lib-generator.
3. IIUC, the DOM lib generator gets its types from `@webref/idl` which, for example, only defines `HTMLScriptElement#src` as a string (https://github.com/w3c/webref/blob/f19b3f5b22dae2b0c91f5209aa351cccce2008a8/ed/idl/html.idl#L1247). It does have definitions for `window.trustedTypes` and some other trusted type APIs though (https://github.com/w3c/webref/blob/main/ed/idl/trusted-types.idl).
4. The status of the spec is ED, so it should be considered in "good standing" by `browser-specs` (https://github.com/w3c/trusted-types/blob/main/spec/index.bs).

It looks like the reason why none of the Trusted Types APIs are showing up right now is because the lib generator has an explicit override to exclude `Trusted*` types, possibly because of inconsistencies. E.g. naively removing that override generates a lot of the API surface but not the `WindowOrWorkerGlobalScope` interface mixin that adds `trustedTypes`.

TL;DR: I'm not sure if it's actually going to just show up at this point since it looks like the MDN + spec preconditions have largely been met but they get excluded by the generator, maybe?

**Comment by jakebailey** [comment]
They won't show up because of that exclusion, no. It could be updated, but we'll have to consider that there are already existing types for Trusted Types out there we probably need to make sure are compatible.

**Comment by hybrist** [comment]
I'm also surprised by some of the inconsistencies, e.g. why some web APIs show the trusted type patches but not `HTMLScriptElement#src`. It's likely confusing if TypeScript enables support before the IDL is actually self-consistent. I asked on the trusted-types spec repo to see if somebody knows what's up with that (https://github.com/w3c/trusted-types/issues/605).
